pkg/migration-tool/reader/path_reader.go

```diff
@@ -0,0 +1,169 @@
+		errChan <- fmt.Errorf("path-read run: prom tsdb querier: %w", err)
+		return
+	}
+	ss := querier.Select(false, nil, labels.MustNewMatcher(labels.MatchRegexp, "", ".*"))
+	type itrMap struct {
+		labels labels.Labels
+		itr    chunkenc.Iterator
+	}
+	var iteratorStore = make(map[string]itrMap)
+	// Fill the iteratorStore with the iterators across all the series-sets for the given time-range.
+	for ss.Next() {
+		series := ss.At()
+		iteratorStore[series.Labels().String()] = itrMap{labels: series.Labels(), itr: series.Iterator()}
+	}
+	go func() {
+		defer func() {
+			if err = tsdb_errors.NewMulti(err, db.Close()).Err(); err != nil {
+				errChan <- fmt.Errorf("path-read run: %w", err)
+			}
+			if err = querier.Close(); err != nil {
+				errChan <- fmt.Errorf("path-read run: %w", err)
+			}
+			close(rr.sigBlockRead)
+			log.Info("msg", "reader is down")
+			close(errChan)
+		}()
+		log.Info("msg", "reader is up")
+		// Start creating blocks and filling them with samples.
+		for rr.plan.ShouldProceed() {
+			select {
+			case <-rr.c.Done():
+				return
+			case <-rr.sigForceStop:
+				return
+			default:
+			}
+			blockRef, err := rr.plan.NextBlock()
+			if err != nil {
+				errChan <- fmt.Errorf("path-run run: %w", err)
+				return
+			}
+			ts := make(tsMap)
+			mint, maxt := blockRef.Mint(), blockRef.Maxt()
+			populateTimeseries := func(m itrMap) bool {
+				// Populate the timeseries map with the iterators for the current time-range delta.
+				// We maintain a local store of series iterators in iteratorStore. This keeps the last progress in the memory
+				// (at a negligible cost since iterator), leading to faster access of samples for subsequent time-ranges.
+				itr := m.itr
+				var samples []prompb.Sample
+				appender := func() bool {
+					t, v := itr.At()
+					if t >= mint && t < maxt {
+						samples = append(samples, prompb.Sample{
+							Timestamp: t,
+							Value:     v,
+						})
+						return true
+					}
+					return false
+				}
+				if itr.Seek(mint) {
+					appender()
+					for itr.Next() {
+						if !appender() {
+							break
+						}
+					}
+				}
+				if len(samples) > 0 {
+					ts[m.labels.String()] = struct {
+						labels  []prompb.Label
+						samples []prompb.Sample
+					}{labels: labelsToPrombLabels(m.labels), samples: samples}
+				}
+				return true
+			}
+			for _, iteratorMap := range iteratorStore {
+				populateTimeseries(iteratorMap)
+			}
+			timeseries, bytesUncompressed := tsMapToTs(ts)
+			ts = nil
+			blockRef.Fill(timeseries, bytesUncompressed)
+			rr.sigBlockRead <- blockRef
+			blockRef = nil
+		}
+	}()
+}
+
+type tsMap map[string]struct {
+	labels  []prompb.Label
+	samples []prompb.Sample
+}
+
+func tsMapToTs(m tsMap) ([]*prompb.TimeSeries, int) {
+	ts := make([]*prompb.TimeSeries, len(m))
+	size := 0
+	i := 0
+	for _, tsm := range m {
+		pts := &prompb.TimeSeries{
+			Labels:  tsm.labels,
+			Samples: tsm.samples,
+		}
+		b, _ := pts.Marshal()
+		size += len(b)
+		ts[i] = pts
```

Comment:
we are marshalling just for a size? That can't be right. Its soo much wasted work.

Reply:
Can we save the byte slice instead of the []prompb.TimeSeries slice? then the work is not wasted.

Reply:
I don't think that is possible, as we read blocks which are in form of iterators than byte slices. Any suggestions?